derive/src/lib.rs

```diff
@@ -126,20 +129,37 @@ fn generate_fields(fields: &FieldsList) -> Vec<TokenStream2> {
             StaticLifetimesReplace.visit_type_mut(&mut ty);
 
             let type_name = clean_type_string(&quote!(#ty).to_string());
-
-            if let Some(i) = ident {
-                quote! {
-                    .field_of::<#ty>(stringify!(#i), #type_name)
-                }
+            let method_call = if is_compact(f) {
+                quote!(.compact_of::<#ty>)
             } else {
-                quote! {
-                    .field_of::<#ty>(#type_name)
-                }
+                quote!(.field_of::<#ty>)
+            };
+            if let Some(ident) = ident {
+                quote!(#method_call(stringify!(#ident), #type_name))
+            } else {
+                quote!(#method_call(#type_name))
             }
         })
         .collect()
 }
 
+/// Look for a `#[codec(compact)]` outer attribute.
+fn is_compact(f: &Field) -> bool {
+    f.attrs.iter().any(|attr| {
+        let mut is_compact = false;
+        if attr.style == AttrStyle::Outer && attr.path.is_ident("codec") {
+            if let Ok(Meta::List(MetaList { nested, .. })) = attr.parse_meta() {
+                if let Some(NestedMeta::Meta(Meta::Path(path))) = nested.iter().next() {
```

Comment:
Good catch, let's do that in the next PR though (i.e. #44 and #45) as we don't really have `skip` here yet.

Reply:
I think it should be in this PR because if `compact` is not in the first position it will be ignored, because AFAICT only the first one is checked via `nested.iter().next()`

Reply:
Fair enough. On it.

Reply:
I'm sure it can wait until Monday morning though :wine_glass:

Reply:
On a second look I think you're actually not reading in the right place. Attribute validation happens [here](https://github.com/paritytech/parity-scale-codec/blob/5593ee726f3f7b03ef20e1bb9e3ae8b9ef029b9e/derive/src/lib.rs#L139), which then calls `check_field_attribute()`, which [enforces a single item](https://github.com/paritytech/parity-scale-codec/blob/5593ee726f3f7b03ef20e1bb9e3ae8b9ef029b9e/derive/src/utils.rs#L202) (the variant equivalent is [here](https://github.com/paritytech/parity-scale-codec/blob/5593ee726f3f7b03ef20e1bb9e3ae8b9ef029b9e/derive/src/utils.rs#L233)).

I'm not sure which attribute combinations would make sense for `parity-scale-codec` to support. `skip` + `compact` not so much, and I'm not sure `compact` + `encoded_as` either. Maybe `index` + `compact`/`encoded_as`? But so far I think they have been pretty strict about attributes as the macro [may panic](https://github.com/paritytech/parity-scale-codec/blob/master/derive/src/utils.rs#L146-L148) otherwise.

The `find_meta_item()` utility function you linked to is general purpose (and totally neat, so I'm cribbing it in #44).

Reply:
You are right it will not allow multiple attributes: https://github.com/paritytech/parity-scale-codec/blob/master/derive/src/utils.rs#L201.

This crosses over into the discussion about whether/how `scale-info` and `parity-scale-codec` can be combined, because we are relying on this attribute validation. So we only become vulnerable here (i.e, the attribute could be ignored) if this validation is relaxed to allow some future valid combination of attributes (I can see now that `#[codec(skip, compact)]` is wrong).

So could either make it future compatible now, or just add a compile test which checks that this validation is in place. Don't mind if you do it here or in one of the follow ups.

Reply:
Yes for now it seems all attribute are exclusive:

For field attribute I think they are all exclusive: either compact, or skip or encoded_as or nothing.
For variant attribute they are also exclusive: either skip or index or nothing.